### ISSUE TITLE
NaN value appears when I calculate a derivative of kernel with respect to relative position

### ISSUE THREAD

**Issue by Nanco-L** [e3nn/e3nn#112]
Hello.

I found that derivative of kernel with respect to relative position shows NaN value in specific case.
(relative position of [0., 0., A]: vector which is parallel to z-axis)

Sample code of above problem is in below:
```python
import torch
from e3nn import Kernel, o3
from e3nn.radial import GaussianRadialModel
from e3nn.non_linearities.rescaled_act import swish
from functools import partial

RadialModel = partial(GaussianRadialModel, max_radius=7.0, number_of_basis=3, h=100, L=1, act=swish)
mykernel = Kernel([(4,0)], [(1,1)], RadialModel=RadialModel)

xx = torch.randn(1,4)
yy = torch.tensor([[0.,0.,2.]], requires_grad=True)

out = torch.einsum('eij,ej->ei', mykernel(yy), xx)
loss = torch.sum(torch.norm(out, dim=1))

dyy = torch.autograd.grad(loss, yy)
print(dyy)

# $ (tensor([[nan, nan, nan]]),)
```

Derivative of network with respect of relative positions of atom is important for calculating atomic force of given structure.
Thus, above phenomenon (or bug?) is very critical for me.
I will try to fix that NaN value for my own but I think developers of e3nn can find the source of this phenomenon faster than me. :)
Please give me some help, thanks!

**Comment by blondegeek** [comment]
I haven't checked this, but try adding an epsilon to your norm
```
eps = 1e-8
loss = torch.sum(torch.norm(out + eps, dim=1))
```

**Comment by blondegeek** [comment]
I also recommend using anomaly detection for figuring out where the NaN is
generated:
https://pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly

**Comment by mariogeiger** [comment]
I guess it's due to the gimbal lock problem because arguments of the spherical harmonics are Euler angles.

**Comment by mariogeiger** [comment]
I think it comes from `atan2` at [this line](https://github.com/e3nn/e3nn/blob/master/e3nn/rsh.py#L253)

```python
import torch
from e3nn import Kernel, o3, rsh

x = torch.tensor([0, 0., 1.], requires_grad=True)

alpha = torch.atan2(x[1], x[0])
alpha.backward()

x.grad
# tensor([nan, nan, 0.])
```

**Comment by Nanco-L** [comment]
Deal all, thanks for fast answer.

First I see that NaN value, I suspect that spherical harmonics might be the source as @mariogeiger mentioned.
I change https://github.com/e3nn/e3nn/blob/86b2bcafe3d6d583b3e71b51effba03ac075ec2e/e3nn/rsh.py#L253 like below.

```python
# alpha = torch.atan2(xyz[..., 1], xyz[..., 0])  # [...] # original line
epsilon = 1e-5
alpha = torch.atan2(xyz[..., 1], xyz[..., 0]+epsilon)
```

And this change removes the NaN value of above sample code.
If that is not the potential source of other problems you think, please apply that.

**Comment by mariogeiger** [comment]
This is not a solution. What happens if `xyz = [0, -epsilon, sqrt(1 - epsilon**2)]` ?

**Comment by Nanco-L** [comment]
Oh, you're right. Then I have no other solution for that problem..
Until above problem is solved, I need to randomly rotate my input with the relative positions of [0.,0.,A].
If anyone find good solution, please notice me!

**Comment by blondegeek** [comment]
@mariogeiger, didn’t we use to have Cartesian spherical harmonics for this
very reason?

**Comment by blondegeek** [comment]
I guess the question is whether this was an existing issue with the old "backwardable" spherical harmonics
https://github.com/mariogeiger/se3cnn/blob/255f1a6694dcaef8bcec0d5daaad8b377cacba3e/se3cnn/SO3.py#L387
I've definitely used these before to backprop to coordinates...

**Comment by mariogeiger** [comment]
It was also using the function `atan2` and evaluating it in `(0, 0)`.
The problem it that this function is not differentiable in `(0, 0)`.
I am afraid that if you don't get a nan it will probably be wrong anyway!

**Comment by blondegeek** [comment]
Looks like I actually did add an eps to atan2 when I was doing this to test
how "conservative" some of our predicted vector fields were.
This is on my `coord_grads` branch of the old `se3cnn`
https://github.com/blondegeek/se3cnn/blob/coord_grads/se3cnn/SO3.py#L186
See the comment added here:
https://github.com/mariogeiger/se3cnn/commit/e35268009a8d9e8ed3f2ea46ed288a778e0c4706

**Comment by blondegeek** [comment]
Closing this since as seen in the previous comments, you can carefully add an epsilon to atan2 if you benchmark it against actual gradients.

**Comment by blondegeek** [comment]
Epsilon added (in multiple places) with https://github.com/e3nn/e3nn/pull/118
See `rsh.py` for detailed comment.
